### PR TITLE
[eme] Add test case for waitingforkey event

### PR DESCRIPTION
**Pull request by mwatson2**

### CODE REVIEW

File: encrypted-media/scripts/playback-temporary-waitingforkey.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            _mediaKeys,
+            _mediaKeySessions = [ ],
+            _mediaSource;
+
+        function onMessage(event) {
+            assert_any( assert_equals, event.target, _mediaKeySessions );
+            assert_true( event instanceof window.MediaKeyMessageEvent );
+            assert_equals( event.type, 'message');
+
+            assert_any( assert_equals,
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: This assert seems unrelated to this test?
General question: Do we want to assert things that are expected to be true but unrelated to the test purpose?

File: encrypted-media/scripts/playback-temporary-waitingforkey.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                        event.messageType,
+                        [ 'license-request', 'individualization-request' ] );
+
+            config.messagehandler( config.keysystem, event.messageType, event.message )
+            .then( function( response ) {
+
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: nit: empty line

File: encrypted-media/scripts/playback-temporary-waitingforkey.js (hunk 1 above)
Comment (mwatson2) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: Yes, you're right. We should probably scrub out asserts which are unrelated to the particular feature being tested.

File: encrypted-media/scripts/playback-temporary-waitingforkey.js (hunk 2 above)
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: There are lots of extra empty lines at the beginning and end of blocks. It seems inconsistent. I think we should remove them.

File: encrypted-media/scripts/playback-temporary-waitingforkey.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            // Not using waitForEventAndRunStep() to avoid too many
+            // EVENT(onTimeUpdate) logs.
+            _video.addEventListener('timeupdate', onTimeupdate, true);
+
+            // Create sessions now if we do not have them
+            if ( _mediaKeySessions.length === 0 ) {
+                config.initData.forEach( function( initData ) {
+
+                    var mediaKeySession = _mediaKeys.createSession( 'temporary' );
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: Spaces inside () are applied inconsistently. For example, here vs. line 42. I prefer no spaces for simplicity and slightly shorter lines, but let's pick one and be consistent.

File: encrypted-media/scripts/playback-temporary-waitingforkey.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        function onWaitingForKey(event) {
+
+            // Not using waitForEventAndRunStep() to avoid too many
+            // EVENT(onTimeUpdate) logs.
+            _video.addEventListener('timeupdate', onTimeupdate, true);
+
+            // Create sessions now if we do not have them
+            if ( _mediaKeySessions.length === 0 ) {
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: We should probably check that we receive the exact number of expected events.

File: encrypted-media/scripts/playback-temporary-waitingforkey.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            return access.createMediaKeys();
+        }).then(function(mediaKeys) {
+            _mediaKeys = mediaKeys;
+
+            _video.setMediaKeys(_mediaKeys);
+
+            waitForEventAndRunStep('waitingforkey', _video, onWaitingForKey, test);
+
+        }).then(function() {
+            return testmediasource(config);
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: This function name is misleading/confusing. Should it be createMediaSourceForConfig?

**Comment by ddorwin** [comment]
I skimmed the JS file and provided comments. Some probably apply more generally to all tests and infrastructure. I'm not familiar enough with the infrastructure to review details, but it seems fine other than the comments.

File: encrypted-media/scripts/playback-temporary-waitingforkey.js (hunk 4 above)
Comment (mwatson2) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: The number and timing of events varied between clearkey and widevine on Chrome. I suspect it may be due to timing. Originally, the test fetched the first key and waiting for playback to start before fetching the second. This worked on clearkey, but with Widevine, it wants both keys before it will begin playback. I guess this could be because the media with the second key is so close to the start (we have 6 seconds with key1 followed by 6 seconds with key 2).

I should probably switch to using the single-key content and testing for a single waitingforkey event before playback starts.

**Comment by mwatson2** [comment]
I filed https://github.com/w3c/web-platform-tests/issues/3423 for cleanup that applies to multiple tests.

File: encrypted-media/scripts/playback-temporary-waitingforkey.js [hunk 6]
```diff
@@ -1,8 +1,12 @@
 function runTest(config) {
 
+    // config.initData contains a list of keys. We expect those to be needed in order and get
+    // one waitingforkey for each one.
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: nit: one waitingforkey **event** for each one

File: encrypted-media/scripts/playback-temporary-waitingforkey.js [hunk 7]
```diff
@@ -16,52 +20,30 @@ function runTest(config) {
...
-
-            assert_any( assert_equals,
-                        event.messageType,
-                        [ 'license-request', 'individualization-request' ] );
+        function onFailure( error ) {
+            forceTestFailureFromPromise(test, error);
+        }
 
+        function onMessage(event) {
+
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: nit: new empty line

File: encrypted-media/scripts/playback-temporary-waitingforkey.js [hunk 8]
```diff
@@ -16,52 +20,30 @@ function runTest(config) {
...
-
-                    _mediaKeySessions.push( mediaKeySession );
-
-                    mediaKeySession.generateRequest( config.initDataType, initData )
-                    .catch(function(error) {
-                        forceTestFailureFromPromise(test, error);
-                    });
-                } );
-            }
+            assert_less_than( _mediaKeySessions.length, config.initData.length );
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: You might explain this since it may not be obvious to future readers.

File: encrypted-media/scripts/playback-temporary-waitingforkey.js [hunk 9]
```diff
@@ -72,19 +54,18 @@ function runTest(config) {
             return access.createMediaKeys();
         }).then(function(mediaKeys) {
             _mediaKeys = mediaKeys;
-
             _video.setMediaKeys(_mediaKeys);
-
             waitForEventAndRunStep('waitingforkey', _video, onWaitingForKey, test);
-
+            
+            // Not using waitForEventAndRunStep() to avoid too many
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: The previous location ensured that the test would never complete without seeing waitingforkey at least one time. The flow is cleaner here, but it could cover up a test or UA bug.

File: encrypted-media/scripts/playback-temporary-waitingforkey.js [hunk 10]
```diff
@@ -16,52 +20,30 @@ function runTest(config) {
...
+        function onMessage(event) {
+        
             config.messagehandler( config.keysystem, event.messageType, event.message )
             .then( function( response ) {
-
                 event.target.update( response )
-                .catch(function(error) {
-                    forceTestFailureFromPromise(test, error);
-                });
+                .catch(onFailure);
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: nite: I find this (lack of) indentation confusing. Elsewhere where we don't have such indentation, the closing parenthesis is immediately before the ')'. I'm not sure that's better. Would just moving this to the line above be better?

ditto on line 42.

**Comment by ddorwin** [comment]
The JS LGTM with comments (from a cursory review as noted previously).

We'll look into the difference in events you noted in https://github.com/w3c/web-platform-tests/pull/3417#discussion_r73228617. If the number and order of events should be identical and we can resolve this difference, it would be nice to eventually restore the use of the `multikey` media since it provides additional coverage.

File: encrypted-media/scripts/playback-temporary-waitingforkey.js (hunk 9 above)
Comment (mwatson2) [inline encrypted-media/scripts/playback-temporary-waitingforkey.js]: I'll add an assert in onTimeupdate to check we have the expected number of sessions.

**Comment by mwatson2** [comment]
@ddorwin Yes, I plan to make another waitingforkey test with multiple in-sequence keys. In particular this would test getting a waitingforkey event during playback, when data encrypted with a different key is encountered.
